### DISCUSSION TITLE
Adding Krux as an option in Sparrow

### DISCUSSION THREAD

**Discussion by craigraw** [selfcustody/krux#253] [General]
Hi team,

I'm the developer for Sparrow Wallet. I'm unfamiliar with the Krux, but I was watching [this guide](https://www.youtube.com/watch?v=JfK2m8ucuxU) indicating how Krux can be used with the SeedSigner and Specter Desktop options for importing the wallet and exporting a multisig config respectively. I have also seen your documentation on using Specter DIY for importing.

Would it be helpful to add Krux as a separate option? If so I'd need a logo and description of the UI flow in the device for each option.

**Comment by tadeubas** [comment]
Yes plz add a separate option for Krux!

@odudex can you check if this info is ok plz?

LOGO:
![logo-krux](https://github.com/selfcustody/krux/assets/5518496/35f17c95-6fb3-4bec-8f50-c6ba7518bea2)

IMPORT > Airgapped Hardware Wallet:
Import file or QR created on your Krux's device by selecting Extended Public Key from the main menu once you have loaded your mnemonic.

EXPORT (this should show an animated QR code, like it does with Jade):
Export a QR that can be read by your Krux's device to import a multisig wallet using the Wallet Descriptor from the main menu once you have loaded your mnemonic.

**Comment by craigraw** [comment]
Thanks. 

Does Krux support `ur:crypto-output` (for the wallet export) by any chance? That would be the most efficient way - the Jade and Specter options use text formats encoded as QRs.

**Comment by tadeubas** [comment]
I can confirm that it works with Specter DIY, Jade, Keystone or Passport QRs...

**Comment by odudex** [comment]
This will be awesome!
I think Krux and Specter are the closest. You can use it as a reference. Krux cameras are low cost so, if possible, would default QR codes densities to medium or low, prioritizing animated QRs instead of high density ones.
Yes, Krux does support UR.
What's the ideal format and resolution (or vector) for the logo image? Transparent background? Or is this @tadeubas sent enough?
To test you could use Krux app if you have an Android phone available, [just install the APK](https://github.com/odudex/krux_binaries/tree/main/Android), it runs the same code as the dedicated devices.
The **UI flow** is straight forward as Krux was made focusing coordinators like Sparrow, just:
- Load a key (Krux will ask if its single or multisig at this step)
- Export the XPUB or ZPUB (as @tadeubas description mention)
 (Krux will export it with header like [3f6e554e/48h/0h/0h/2h]Zpub75utHse...)
- Load it in Sparrow.
- If singlesig it will be already possible to verify the addresses, sign transactions
- If multisig, export QR descriptor (like Specter) on Sparrow, import it on Krux (by clicking "Wallet Descriptor"), then addresses verification and and signatures will be available.

**Comment by odudex** [comment]
I just added some more [logo options here](https://github.com/odudex/krux_binaries/tree/main/branding)

**Comment by craigraw** [comment]
Thanks for the info - I'll try it on an Android. The images shared so far are fine too. Quick question:

Does Krux scan the QR shown when clicking the QR display button to the right of the descriptor field on the Settings tab in Sparrow? It's a `ur:crypto-output` export of the current wallet.

**Reply by odudex** [reply]
> Does Krux scan the QR shown when clicking the QR display button to the right of the descriptor field on the Settings tab in Sparrow? It's a `ur:crypto-output` export of the current wallet.

Yes, it scans and loads the descriptor correctly.

**Reply by craigraw** [reply]
Ok great - then it's the same as SeedSigner, which does not require a specific wallet export option as the normal QR export works. I'll still add the import option with Krux as another wallet type.

**Reply by ghost** [reply]
Thank you @craigraw! I know this is just another in a long line of devices you've added support for, but this is quite a milestone for Krux and personally very fulfilling to see happen :) And while you're here, thank you for the great OSS work you've done with Sparrow.

**Comment by craigraw** [comment]
Krux has been added to Sparrow as an airgapped hardware wallet, and appears in the list of airgapped import options:

<img width="1209" alt="Screenshot 2023-10-05 at 10 37 21" src="https://github.com/selfcustody/krux/assets/862166/d29e047d-0ef5-41ae-9339-e36ea1ba1ff5">

Two questions that came up:
1. Is import by file an option? It was included in the import instructions above, but I don't see any way to export a file using the Extended Public Key export in the Android app. I've excluded it for now.
2. I can't seem to get message signing to return a valid signature. Sparrow has recently implemented (not yet released) message signing by QR code as per the Specter method and recently implemented by SeedSigner, as documented here: https://github.com/SeedSigner/seedsigner/pull/262#issue-1419886558. It appears the public key derived by Krux does not match the one in Sparrow.

I've committed the initial Krux integration, so you can test it by running from source, which is fairly simple and described here: https://github.com/sparrowwallet/sparrow

**Comment by tadeubas** [comment]
@craigraw Unfortunately the Android version cannot use SD card, so it doesn't show the option to save XPUB/ZPUB to SD card, but if you use the simulator with the `--sd` option or a device with a compatible SD card the option will appear

**Comment by odudex** [comment]
Message signing workflow on Krux is straight forward.
- If Krux receives a hash, it will sign it. If it is a string, will hash then sign.
- The public key it exports is raw hex. Which can even be converted to PEM so verification can be done with openssl.
We wrote some [simple scripts](https://github.com/odudex/krux-file-signer) to help use Krux to sign files, create PEM pubkeys and verify signatures. They're also a good way to check the workflow we use.
I'll check the Seedsigner discussion to try to get their method.

**Comment by odudex** [comment]
Ran it from source, everything worked as expected! Fantastic!

**Comment by craigraw** [comment]
Great, thanks for testing :)

> I'll check the Seedsigner discussion to try to get their method.

I think the advantage (if I understand your explanation) is the specification of a derivation path allows signing using a specified address.

**Reply by odudex** [reply]
Yes, it seems the difference is Krux won't use a derived address like the others.
@jeff, would we still be able to sign files and use openSSL for verification if address derived public keys were used?
@craigraw can you elaborate on the benefits this extra step/field brings? Why would I want to sign with a specific address?

**Reply by craigraw** [reply]
> can you elaborate on the benefits this extra step/field brings? Why would I want to sign with a specific address?

The immediate advantage is you can prove ownership of a particular address. But an address is a useful handle for a public key in a wallet, which is in turn a useful grouping of addresses, so there is a practical aspect of using different addresses in a wallet for different signing/verification purposes.

**Reply by odudex** [reply]
@jreesun how about we add a 3rd item to the sign menu?
Message (BIP322)

**Reply by craigraw** [reply]
Just a note -  BIP322 may be the future of message signing, but it's not well supported - in fact Sparrow's implementation is the only one in a wallet I'm aware of. It's also more complex to implement than legacy message signing, fwiw, even the "simple" approach.

**Reply by tadeubas** [reply]
Maybe I am missing something, but why did Specter, SeedSigner and Sparrow implemented this? It is used by some service? It was demanded by the users to a real use? IMHO there are other great functionalities to be done rather than just supporting a random BIP

**Reply by odudex** [reply]
Sure, I think utility should be priority, interoperability is also important and simplicity should be in mind too. A well though BIP that meets the prior items would be handy. To be honest, the only utility I used our message signing until now was to sign files, (awesome utility by the way :smile: )
But sure there are other plans for message signing I'm not aware of, like apps and protocols. So folks, please share the utilities you know of.
Trying to predict some possibilities, I even think about an air-gapped encrypted messages protocol, making use of signing + encryption. We have all the blocks required to build this.

**Reply by tadeubas** [reply]
Sure, the signing for Nostr would be awesome!

**Comment by odudex** [comment]
@craigraw could you update your [Krux app](https://github.com/odudex/krux_binaries/tree/main/Android) and test if message signing behaves as you expected? I adapted some SpecterDIY code to Krux to make it compatible.
If it is, I'll add this to all devices. I did it in a way our previous method of signing with master, which we also use to sign our binaries will still work as before.

**Comment by craigraw** [comment]
Tested, works correctly!
